cdptools/databases/document_store_database.py

```diff
@@ -0,0 +1,840 @@
+            return filt
+        elif isinstance(filt, (list, tuple)):
+            # Assume equal
+            if len(filt) == 2:
+                return WhereCondition(filt[0], WhereOperators.eq, filt[1])
+            elif len(filt) == 3:
+                return WhereCondition(*filt)
+            else:
+                raise exceptions.UnstructuredWhereConditionError(filt)
+        else:
+            raise exceptions.UnknownTypeWhereConditionError(filt)
+
+    @staticmethod
+    def _construct_orderby_condition(by: Union[OrderCondition, List, Tuple, str]):
+        """
+        Construct an orderby condition from the passed by object.
+
+        For available order condition operators, please view the OrderOperators class 
+        attributes.
+
+        Parameters
+        ----------
+        by: Union[OrderCondition, List, Tuple, str]
+            The order by data to construct an OrderCondition for.
+
+        Returns
+        -------
+        by: OrderCondition
+            If provided an already constructed OrderCondition, simply 
+            returns the provided object.
+            If provided a string, returns and OrderCondition with the provided 
+            string being used as the column name and defaults to descending as the 
+            operator.
+            If provided a List or Tuple of length two, constructs an OrderCondition 
+            with the provided data.
+            Provided List or Tuple objects should follow the format:
+                {target_to_order}, {operator}
+
+        Examples
+        --------
+        ```
+        condition = Database._construct_orderby_condition(OrderCondition(
+            column_name="event_id",
+            operator="DESCENDING",
+        ))
+        # Returns the same OrderCondition that was passed in
+
+        condition = Database._construct_orderby_condition("event_id")
+        # Returns the same OrderCondition as above
+
+        condition = Database._construct_orderby_condition(["event_id", 
+                                                          OrderOperators.desc])
+        # Returns the same OrderCondition as above
+
+        condition = Database._construct_where_condition(("event_id", "ASCENDING")
+        # Returns an OrderCondition similar to above but the operator is set to 
+        ascending rather than descending.
+        ```
+        """
+        if isinstance(by, OrderCondition):
+            return by
+        if isinstance(by, str):
+            # Assume descending
+            return OrderCondition(by, OrderOperators.desc)
+        elif isinstance(by, (list, tuple)):
+            # Assume descending
+            if len(by) == 1:
+                return OrderCondition(by[0], OrderOperators.desc)
+            elif len(by) == 2:
+                return OrderCondition(*by)
+            else:
+                raise exceptions.UnstructuredOrderConditionError(by)
+        else:
+            raise exceptions.UnknownTypeOrderConditionError(by)
```

Comment:
Same mental note here. I think the long term plan will likely be to have these plus the above extra objects in a `database.py` file. Then have this `DocumentStoreDatabase` inherit from `Database`. Just more levels of abstraction.

I say mental note but I may talk to you and To about this tonight. It may be good to make a `release-3` branch where all of this is done and we can kill off the old stuff that we are replacing.

Reply:
Even more thinking. If we do make a `release-3` branch. Would the ideal abstract route be:

`Database` -> `DocumentStore` -> `CloudFirestore` or would this entire class just replace the current `Database` and have this be `Database` -> `CloudFirestore` :thinking:

Reply:
I agree if we are going to make a distinction between relational and non-relational databases, I think we should use the correct terminologies (tables and rows vs collections and documents).

I like the `DocumentStoreDatabase` idea. Maybe have all the `get_or_upload_X` in `Database`. And in `DocumentStoreDatabase` we'll have the appropriate document store methods like `select_document_by_id` instead of `select_row_by_id`?

Or if we could think of a more generic name that is a level above rows and documents (and a generic name for tables and collections too), we can have `select_GENERIC_NAME_by_id` in `Database` too. Though I can't think of a good generic name.

Reply:
I can change it to be `collections` and `documents` for now, and if we decide to make it more generic with another layer I don't think that'd be a big refactor.

I think adding another layer of abstraction like `Database` -> `DocumentStore` -> `CloudFirestore` makes sense if we want/plan to have a relational and non-relational databases, since we will have universal methods like the reshape and select methods. Since relational vs non-relational could have different tables/collections and structure, I think the intermediate level would be beneficial.

But If we're only planning on using a non-relational database, I don't really see a point in having 3 layers.

Reply:
Just had the front-end meeting and I think we are going to settle on the 3 level option.


```
            > DocumentStoreDatabase -> CloudFirestore
          /
Database -
          \
           > RelationalDatabase
```

It was asked by a front-end dev if the system is directly tied to a firestore instance or if it can work on any datastore and while I think it was a question out of curiosity, it pushed me towards the 3 level model.